### ISSUE TITLE
Performance Improvement Idea for AJAX Spider (Don't request URLs already Spidered)

### ISSUE THREAD

**Issue by davewichers** [zaproxy/zaproxy#6936]
I'm crawling an OWASP Benchmark style app I'm developing, and I notice that the AJAX spider is requesting copies of each .css, .js, .png, etc. for EVERY page it crawls, even though those resources are the same for every page.

Seems like the AJAX Spider should cache all the URLs it has requested during a crawl so far, and check to see if the URL it found is unique, before requesting it. I think that would make the AJAX spider much faster for certain apps. Admittedly, OWASP Benchmark style apps have ALOT of repetition, so this might not be as much of a problem for normal apps, but I still think it would improve performance some.

Not critical. Just an idea.  The same idea might also apply to the normal spider. Not sure.

**Comment by thc202** [comment]
It's actually the browser that requests those resources, are they being cached (when not using the AJAX spider)?

**Comment by davewichers** [comment]
I'm using the AJAX Spider within ZAP itself (Attack-->AJAX Spider). How does that involve the browser? And yes, all these duplicate requests return 304 Not Modifieds.

**Comment by kingthorin** [comment]
The Ajax Spider launches headless browsers to do its work.

**Comment by davewichers** [comment]
OK. But can't it control the browser behavior so it doesn't issue repetitive requests?

**Comment by psiinon** [comment]
The AJAX spider doesnt "know" what requests the browser will make. It only "knows" about things like links and HTML input elements. It clicks on those and fills out forms, the browser then does whatever those controls tell it to do.
Automating a browser to efficiently crawl one specific site is typically not too hard a task, but its a custom approach.
But thats not we need to do - we need to do our best to make sure that the AJAX spider crawls all sites effectively. Thats hard.

The most thorough way to explore an app is for a human to control the browser, but thats no good for automation.
The most efficient way to explore an app is to import a complete definition of that site, eg using OpenAPI
An effective way to explore an app is by proxying thorough unit/integrations tests through ZAP.
If your site does not rely on JavaScript links then the traditional spider is a good way to explore it.

If you dont make your app easy to explore then you will have to use less efficient options, like the AJAX spider :)

**Comment by davewichers** [comment]
Simon, thanks for the explanation. And while I understand it, to 'make
sure' ZAP does the best it can, I routinely run both the regular spider and
then the AJAX spider to make sure ZAP has the best info available before
running the scan. As such, I think making the AJAX spider faster would be
good for everyone. Even when people are running it that don't need to.
Because that doesn't actually matter if they are running it anyway.

**Comment by psiinon** [comment]
Oh I'm all for making it faster :)
But not at the expense of missing things (on the default settings) - we've had more complaints that it does not find enough :/

**Comment by Jeeppler** [comment]
@davewichers which browser/webdriver do you use with ZAP? 

You mentioned you see a lot 304 not modified responses, is that correct?

**Comment by kingthorin** [comment]
304 are loaded from cache (generally).

**Comment by Jeeppler** [comment]
@kingthorin yes exactly, those are loaded from cache. However, one can disable the cache. I was just wondering whether the issue of @davewichers is related to a disabled cache or a similar browser setting.

**Comment by kingthorin** [comment]
Possible, we’ll have to check. Kinda depends where he is seeing them as 304s.

**Comment by davewichers** [comment]
I'm using the default FireFox built into ZAP with no config changes.

**Comment by psiinon** [comment]
Closing as we are now focussing on the Client Spider.
If this issue still occurs with the Client Spider then please raise a new issue.
